llvm/lib/CodeGen/GlobalISel/Utils.cpp

```diff
@@ -660,8 +660,17 @@ std::optional<APInt> llvm::ConstantFoldBinOp(unsigned Opcode,
   default:
     break;
   case TargetOpcode::G_ADD:
-  case TargetOpcode::G_PTR_ADD:
     return C1 + C2;
+  case TargetOpcode::G_PTR_ADD: {
+    // Types can be of different width here.
+    if (C1.getBitWidth() > C2.getBitWidth())
+      return C2.zext(C1.getBitWidth()) + C1;
+    // We always need to return something the same size as C1, so
+    // truncate in this case.
+    if (C1.getBitWidth() < C2.getBitWidth())
+      return C1 + C2.trunc(C1.getBitWidth());
+    return C1 + C2;
```

Comment:
Just use C2.zextOrTrunc unconditionally? Also, why is zero extending correct? (Why not sign extending for example?)

Reply:
I think sext is probably more correct as offsets can be negative, right?

Reply:
I think SEXT is the correct one. It's what the legalizer does for widenScalar on `G_PTR_ADD`.

Reply:
OK.